File: edukate-backend/src/main/java/io/github/sanyavertolet/edukate/backend/entities/Bundle.java
package io.github.sanyavertolet.edukate.backend.entities;

import io.github.sanyavertolet.edukate.backend.dtos.BundleDto;
import io.github.sanyavertolet.edukate.backend.dtos.BundleMetadata;
import io.github.sanyavertolet.edukate.backend.dtos.CreateBundleRequest;
import io.github.sanyavertolet.edukate.common.Role;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.With;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.PersistenceCreator;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.*;

@Data
@AllArgsConstructor(onConstructor = @__(@PersistenceCreator))
@RequiredArgsConstructor
@Document(value = "bundles")
public class Bundle {
    @Id
    private String id;
    private String name;
    private String description;
    private Boolean isPublic;

    private List<String> problemIds;
    private Map<String, Role> userIdRoleMap;
    private Set<String> invitedUserIds;

    @With
    @Indexed(unique = true)
    private String shareCode = null;

    @SuppressWarnings("UnusedReturnValue")
    public boolean addUser(String userId, Role role) {
        return userIdRoleMap.putIfAbsent(userId, role) == null;
    }

    @SuppressWarnings("UnusedReturnValue")
    public boolean changeUserRole(String userId, Role newRole) {
        if (!userIdRoleMap.containsKey(userId)) {
            return false;
        }
        userIdRoleMap.put(userId, newRole);
        return true;
    }

    public Role getUserRole(String userId) {
        return userIdRoleMap.getOrDefault(userId, null);
    }

    public boolean isUserInBundle(String userId) {
        return getUserRole(userId) != null;
    }

    public boolean isUserInvited(String userId) {
        return invitedUserIds.contains(userId);
    }

    @SuppressWarnings("UnusedReturnValue")
    public boolean inviteUser(String userId) {
        return invitedUserIds.add(userId);
    }

    public boolean removeInvitedUser(String userId) {
        return invitedUserIds.remove(userId);
    }

    public boolean isAdmin(String userId) {
        return Role.ADMIN.equals(getUserRole(userId));
    }

    @SuppressWarnings("UnusedReturnValue")
    public boolean removeUser(String userId) {
        return userIdRoleMap.remove(userId) != null;
    }

    public List<String> getAdminIds() {
        return userIdRoleMap.entrySet().stream().filter(entry -> entry.getValue().equals(Role.ADMIN))
                .map(Map.Entry::getKey).toList();
    }

    public static Bundle fromCreateRequest(CreateBundleRequest bundleRequest, String adminId) {
        return new Bundle(
                null,
                bundleRequest.getName(),
                bundleRequest.getDescription(),
                Boolean.TRUE.equals(bundleRequest.getIsPublic()),
                new ArrayList<>(bundleRequest.getProblemIds() != null ? bundleRequest.getProblemIds() : List.of()),
                new HashMap<>(Map.of(adminId, Role.ADMIN)),
                new HashSet<>(),
                null
        );
    }

    public BundleDto toDto() {
        return new BundleDto(name, description, null, isPublic, null, shareCode);
    }

    public BundleMetadata toBundleMetadata() {
        return new BundleMetadata(name, description, null, shareCode, isPublic, (long) problemIds.size());
    }
}

File: edukate-backend/src/main/kotlin/io/github/sanyavertolet/edukate/backend/entities/Bundle.kt
package io.github.sanyavertolet.edukate.backend.entities

import io.github.sanyavertolet.edukate.backend.dtos.BundleDto
import io.github.sanyavertolet.edukate.backend.dtos.BundleMetadata
import io.github.sanyavertolet.edukate.backend.dtos.CreateBundleRequest
import io.github.sanyavertolet.edukate.backend.dtos.ProblemMetadata
import io.github.sanyavertolet.edukate.common.users.UserRole
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.index.Indexed
import org.springframework.data.mongodb.core.mapping.Document

@Document("bundles")
data class Bundle(
    @field:Id
    val id: String? = null,
    val name: String,
    val description: String,
    val isPublic: Boolean,
    val problemIds: List<String>,
    val userIdRoleMap: Map<String, UserRole>,
    val invitedUserIds: Set<String> = emptySet(),
    @field:Indexed(unique = true)
    val shareCode: String,
) {
    fun getUserRole(userId: String?): UserRole? = userIdRoleMap[userId]

    fun isUserInBundle(userId: String?): Boolean = getUserRole(userId) != null

    fun isUserInvited(userId: String?): Boolean = userId in invitedUserIds

    fun isAdmin(userId: String): Boolean = UserRole.ADMIN == getUserRole(userId)

    fun getAdminIds(): List<String> = userIdRoleMap.filterValues { it == UserRole.ADMIN }.keys.toList()

    fun withJoinedUser(userId: String, role: UserRole): Bundle = copy(
        userIdRoleMap = userIdRoleMap + (userId to role),
        invitedUserIds = invitedUserIds - userId,
    )

    fun withUserRole(userId: String, role: UserRole): Bundle = copy(userIdRoleMap = userIdRoleMap + (userId to role))

    fun withInvitedUser(userId: String): Bundle = copy(invitedUserIds = invitedUserIds + userId)

    fun withoutInvitedUser(userId: String): Bundle = copy(invitedUserIds = invitedUserIds - userId)

    fun withoutUser(userId: String): Bundle = copy(userIdRoleMap = userIdRoleMap - userId)

    fun withVisibility(isPublic: Boolean): Bundle = copy(isPublic = isPublic)

    fun withProblemIds(problemIds: List<String>): Bundle = copy(problemIds = problemIds.toList())

    fun toDto(problems: List<ProblemMetadata>, admins: List<String>): BundleDto = BundleDto(
        name = name,
        description = description,
        isPublic = isPublic,
        shareCode = shareCode,
        admins = admins,
        problems = problems,
    )

    fun toBundleMetadata(admins: List<String>): BundleMetadata = BundleMetadata(
        name = name,
        description = description,
        shareCode = shareCode,
        isPublic = isPublic,
        size = problemIds.size.toLong(),
        admins = admins
    )

    companion object {
        @JvmStatic
        fun fromCreateRequest(bundleRequest: CreateBundleRequest, adminId: String, shareCode: String) = Bundle(
            name = bundleRequest.name,
            description = bundleRequest.description,
            isPublic = bundleRequest.isPublic,
            problemIds = bundleRequest.problemIds.toList(),
            userIdRoleMap = mapOf(adminId to UserRole.ADMIN),
            shareCode = shareCode,
        )
    }
}
